### DISCUSSION TITLE
query_expression values lost after performing union query

### DISCUSSION THREAD

**Discussion by Jackevansevo** [sqlalchemy/sqlalchemy#8879] [Usage Questions] [answered]
Reaching out for some help here as I've been pretty stumped by this recently. Would appreciate any pointers.

I've reduced my issue down to the simplest reproducible example I could. 

You can run a Replit version of the script below interactively here: https://replit.com/@jackevansevo/LightblueDarkgreenSystems#main.py

Ideally I'd expect the output of this script to be:

```
True
False
[True, False]
```

However, my script is printing

```
True
False,
[None, None]
```

```python
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import literal
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy import union_all
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import query_expression
from sqlalchemy.orm import Session
from sqlalchemy.orm import with_expression



Base = declarative_base()


class Tweet(Base):
    __tablename__ = "tweet"
    
    id = Column(Integer, primary_key=True)
    content = Column(String)
    liked = query_expression()


engine = create_engine("sqlite+pysqlite:///")
Base.metadata.create_all(engine)


with Session(engine) as session:

    tweet_a = Tweet(content="A")
    tweet_b = Tweet(content="B")

    session.add_all([tweet_a, tweet_b])

    session.commit()

# Commenting out this line fixes the query...
with Session(engine) as session:
    query_a = (
        select(Tweet)
        .options(with_expression(Tweet.liked, literal(True)))
        .where(Tweet.id == 1)
    )

    print(session.scalars(query_a).first().liked)

    query_b = (
        select(Tweet)
        .options(with_expression(Tweet.liked, literal(False)))
        .where(Tweet.id == 2)
    )

    print(session.scalars(query_b).first().liked)

    union_query = select(Tweet).from_statement(union_all(query_a, query_b))

    print([t.liked for t in session.scalars(union_query).all()])
```

The first query here prints `True` (as I'd expect)

```python
    query_a = (
        select(Tweet)
        .options(with_expression(Tweet.liked, literal(True)))
        .where(Tweet.id == 1)
    )

    print(session.scalars(query_a).first().liked)
```

The second query here prints `False` (as I'd expect)

```python
    query_b = (
        select(Tweet)
        .options(with_expression(Tweet.liked, literal(False)))
        .where(Tweet.id == 2)
    )

    print(session.scalars(query_b).first().liked)
```

However, when I perform a Union of the two queries, I end up with `[None, None]`

```python
    union_query = select(Tweet).from_statement(union_all(query_a, query_b))

    print([t.liked for t in session.scalars(union_query).all()])
```

 Is there anyway to preserve the value of `Tweet.liked` after the two queries have been joined?

Observation: What's intriguing is that the final union query appears to succeed if I run it as part of the same session (as where I create the data), but only if the original two queries have already been executed.

I have a feeling this is something to do with the way sessions in SQLAlchemy behave, but I can't root out the error in my program.

**Comment by zzzeek** [comment]
whoops, you're using from_Statement, ignore my answer here

**Reply by zzzeek** [reply]
you see the True, False when you dont make a new session because it's returning the cached value from the Session where those two attributes are already set.

otherwise this is probably partially a bug - you should have the option in your final select like this:

```py
    uq = union_all(query_a, query_b)
    union_query = (
        select(Tweet)
        .from_statement(uq)
        .options(with_expression(Tweet.liked, uq.selected_columns.liked))
    )

```

but if you add that, it still doesnt work. I dont have a fix right now.  There should be workarounds, but they are failing also unfortunately.   So I dont have this query for you right now.  See #8881

**Reply by zzzeek** [reply]
this is the worst case workaround, and with_expression() is blowing up.

```py
    uq = union_all(query_a, query_b).subquery()

    tq = aliased(Tweet, uq)

    for tweet in session.scalars(
        select(tq)
        .options(with_expression(tq.liked, uq.c.liked))
    ):
        print(tweet.liked)


```

**Reply by zzzeek** [reply]
OK good news, workaround form *does* work, it's because I named the column the same thing...WHEW

so worst case workaround works, start w./ that:

```py
with Session(engine) as session:

    query_a = select(Tweet, literal(True).label("foo")).where(Tweet.id == 1)

    query_b = select(Tweet, literal(False).label("foo")).where(Tweet.id == 2)

    uq = union_all(query_a, query_b).subquery()

    tq = aliased(Tweet, uq)

    for tweet in session.scalars(
        select(tq)
        .options(with_expression(tq.liked, uq.c.foo))
    ):
        print(tweet.liked)


```

from_Statement form (yes! also works, WHEW.   so weird)

```py
with Session(engine) as session:

    query_a = select(Tweet, literal(True).label("foo")).where(Tweet.id == 1)

    query_b = select(Tweet, literal(False).label("foo")).where(Tweet.id == 2)

    uq = union_all(query_a, query_b)

    for tweet in session.scalars(
        select(Tweet).from_statement(uq)
        .options(with_expression(Tweet.liked, uq.selected_columns.foo))
    ):
        print(tweet.liked)


```

OK!  so we can start with that

**Reply by zzzeek** [reply]
that is, when I used `label('liked')`, that was causing the problem.

**Comment by zzzeek** [comment, marked as answer]
OK here is the answer!   with_expression() is a column loader option and therefore it does not take effect inside of subqueries.   So the expected way for this to work is to put with_expression() only at the top level.  for subqueries, you need to use regular Core style approaches to render the expression inside. Make sure not to use label names that match the attribute on the class though ! :)  will see if i can understand that part better and maybe see if it can be fixed.


```py
with Session(engine) as session:
    query_a = select(Tweet, literal(True).label("l")).where(Tweet.id == 1)

    query_b = select(Tweet, literal(False).label("l")).where(Tweet.id == 2)

    uc = union_all(query_a, query_b)
    union_query = (
        select(Tweet)
        .from_statement(uc)
        .options(with_expression(Tweet.liked, uc.selected_columns.l))
    )

    print([t.liked for t in session.scalars(union_query).all()])

```

**Reply by Jackevansevo** [reply]
Thanks so much for such a rapid and detailed response, this has been extremely helpful. I believe this resolves most of what I'm trying to achieve. 

I gave this a go this morning and quickly ran into a situation where both queries contain same overlapping `Tweet` object, i.e:

```python
    query_a = select(Tweet, literal(True).label("l")).where(Tweet.id == 1)
    query_b = select(Tweet, literal(False).label("l")).where(Tweet.id == 1)

    uc = union_all(query_a, query_b)
    union_query = (
        select(Tweet)
        .from_statement(uc)
        .options(with_expression(Tweet.liked, uc.selected_columns.l))
    )

    print([t.liked for t in session.scalars(union_query).all()])
```

Intuitively I'd expect this to be `[True, False]` but in reality the union query ends up returning `[True, True]`.

I'm guessing this is because of the caveats associated with `with_expression` here: https://docs.sqlalchemy.org/en/14/orm/mapped_sql_expr.html#query-time-sql-expressions-as-mapped-attributes

Even when I resort to `populate_existing`, I end up getting the cached result of whichever query ran first. So I'm guessing this is intended behavior (where it's caching the result on a per-object basis).

```python
uc = union_all(query_a, query_b)
union_query = (
    select(Tweet)
    .from_statement(uc)
    .options(with_expression(Tweet.liked, uc.selected_columns.l))
    .execution_options(populate_existing=True)
)
```

Ideally I'd like a way to preserve a label from each query to denote which query the object originated from. I.e. this tweet is a `retweet=True` because it came from the `retweets` query, whereas this same instance of the tweet is `retweeted=False` because it came from another query.

**Reply by zzzeek** [reply]
> Thanks so much for such a rapid and detailed response, this has been extremely helpful. I believe this resolves most of what I'm trying to achieve.
> 
> I gave this a go this morning and quickly ran into a situation where both queries contain same overlapping `Tweet` object, i.e:
> 
> ```python
>     query_a = select(Tweet, literal(True).label("l")).where(Tweet.id == 1)
>     query_b = select(Tweet, literal(False).label("l")).where(Tweet.id == 1)
> 
>     uc = union_all(query_a, query_b)
>     union_query = (
>         select(Tweet)
>         .from_statement(uc)
>         .options(with_expression(Tweet.liked, uc.selected_columns.l))
>     )
> 
>     print([t.liked for t in session.scalars(union_query).all()])
> ```
> 
> Intuitively I'd expect this to be `[True, False]` but in reality the union query ends up returning `[True, True]`.

OK well that is exactly because you're using with_expression(), which is going to populate the entry on the Tweet object, which is unique on identity.  that's the identity map.

if you want Tweet objects and a separate field that varies independently of the identifier of the Tweet, then just select that

```py
uc = union_all(query_a, query_b)
union_query = (
    select(Tweet, uc.selected_columns.l)
    .from_statement(uc)
)
```

> 
> Even when I resort to `populate_existing`, I end up getting the cached result of whichever query ran first. So I'm guessing this is intended behavior (where it's caching the result on a per-object basis).

yeah once the ORM delivers a Tweet with a certain identity, that's the object.  subsequent rows with the same identity are discarded.

> 
> ```python
> uc = union_all(query_a, query_b)
> union_query = (
>     select(Tweet)
>     .from_statement(uc)
>     .options(with_expression(Tweet.liked, uc.selected_columns.l))
>     .execution_options(populate_existing=True)
> )
> ```
> 
> Ideally I'd like a way to preserve a label from each query to denote which query the object originated from. I.e. this tweet is a `retweet=True` because it came from the `retweets` query, whereas this same instance of the tweet is `retweeted=False` because it came from another query.

you would need to keep the Tweet object, which is unique on identity, and the other information you want about a particular context, as separate fields in your row.
